### PR TITLE
PR: Rename rules.constraints to rules.exclude #229

### PR DESCRIPTION
**Pull request by JamesBondOOO7**
Hey @ketkarameya,
I have made the required changes.

Kindly review the PR.

Update:

**Background and Problem**: 
- Renaming the `constraint` attribute to `filter/exclude`. 
- Reference : [review](https://github.com/uber/piranha/pull/225/files/ea753b3e7a7fd7d6890674a0b789d80e61f992b0#r959090037)

**Solution**:
- Changing the Constraint struct to the Filter struct
- Changing constraints attribute with filters

Test Strategy :
- Leaving some `toml` files unchanged so as to test for backward compatibility

Related Github Issue : 
- [Issue_229](https://github.com/uber/piranha/issues/229)

### CODE REVIEW

File: src/models/exclude_filter.rs [hunk 1]
```diff
@@ -15,8 +15,8 @@ use getset::Getters;
 use serde_derive::Deserialize;
 
 #[derive(Deserialize, Debug, Clone, Hash, PartialEq, Eq, Getters)]
-pub(crate) struct Constraint {
-  /// Scope in which the constraint query has to be applied
+pub(crate) struct ExcludeFilter {
```
Comment (ketkarameya) [inline src/models/exclude_filter.rs:18]: Let's rename this struct.
```suggestion
pub(crate) struct Filter {
```

File: src/models/exclude_filter.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ pub(crate) struct Constraint {
   queries: Vec<String>,
```
Comment (ketkarameya) [inline src/models/exclude_filter.rs:25]:
```suggestion
  exclude: Vec<String>,
```
Let's rename `queries` to `exclude`, what do you think? 

I was thinking later when we want to implement #309, we can add a field called `includes` and implement its semantics.

File: src/models/exclude_filter.rs [hunk 3]
```diff
@@ -15,8 +15,8 @@ use getset::Getters;
 use serde_derive::Deserialize;
```
Comment (ketkarameya) [inline src/models/exclude_filter.rs:16]: On the same lines, let's rename `constraints.rs` to `filters.rs`

File: src/models/rule.rs [hunk 4]
```diff
@@ -58,10 +58,10 @@ pub(crate) struct Rule {
   #[serde(default)]
   #[get = "pub"]
   holes: HashSet<String>,
-  /// Additional constraints for matching the rule
+  /// Additional filters for matching the rule
   #[serde(default)]
   #[get = "pub"]
```
Comment (ketkarameya) [inline src/models/rule.rs:63]: I think you forgot to use the alias trick.

**Review by ketkarameya — CHANGES_REQUESTED** [review]
Thanks a lot for this contribution.
After seeing the changes it got me thinking. I think it would be better to `Constraints` -> `Filters` and `Constraints.queries` -> `Filters.exclude`. I think it will make the semantics of the filter even more clear. 
I think you have forgotten to add the `alias` annotation to renamed field in `rule.rs`. 
Also can keep one of the `toml` cases as `rules.constraints` (this will help us test backward compatibility)
You might have to add the same annotation if you decide to rename `queries` to `exclude`.

Also, can you make the PR description more descriptive? You can add things in the following format: 
**Background :**
**Problem :**
**Solution :**
**Test Strategy :**
**Related Github Issue :**
It does not have to be long and descriptive. It should be just enough for anyone in the future to understand why and what changes were applied.

**Review by ketkarameya — COMMENTED** [review]
Took another stab. Just using this opportunity to improve the documentation.

File: src/models/exclude_filter.rs (hunk 3 above)
Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs:16]: It's named `constraint.rs`. So I think it should be renamed to `filter.rs`.
Or shall I keep it as `filters.rs`?

Comment (ketkarameya) [reply src/models/exclude_filter.rs:16]: `filter.rs`

Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs]: Updated accordingly

File: src/models/exclude_filter.rs (hunk 1 above)
Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs]: Updated accordingly

File: src/models/rule.rs (hunk 4 above)
Comment (JamesBondOOO7) [reply src/models/rule.rs:63]: Updated accordingly

File: src/models/exclude_filter.rs [hunk 5]
```diff
@@ -15,8 +15,8 @@ use getset::Getters;
...
 
 #[derive(Deserialize, Debug, Clone, Hash, PartialEq, Eq, Getters)]
```
[the remark this answers is not in the document]
Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs]: Added the comments

File: src/models/exclude_filter.rs [hunk 6]
```diff
@@ -15,8 +15,8 @@ use getset::Getters;
...
-pub(crate) struct Constraint {
-  /// Scope in which the constraint query has to be applied
+pub(crate) struct ExcludeFilter {
+  /// Scope in which the filter query has to be applied
```
[the remark this answers is not in the document]
Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs]: Added the comments

File: src/models/exclude_filter.rs (hunk 2 above)
Comment (JamesBondOOO7) [reply src/models/exclude_filter.rs]: We could do that, or we can also address this once we start implementing `includes`?

**Comment by ketkarameya** [comment]
can you link me to the test case where you are testing backward compatibility ?

**Comment by JamesBondOOO7** [comment]
[Here](https://github.com/uber/piranha/pull/352/commits/c0e3fa6223e601dc36f3ba1c4624cbcb829216dd) is the commit where I used `constraints` instead of `filters` in the `toml` file.

**Comment by ketkarameya** [comment]
Ow! I see you did that backward incompatibility test inside our core cleanup rules :) Can we use the latest name (`filter`) for the core cleanup rules ? 
Maybe u can undo the changes u in the ^ commit and rather make similar change for a test case?

**Comment by JamesBondOOO7** [comment]
Sure, I will undo them.
Also, can I test `filter` -> `constraint` inside this [dir](https://github.com/uber/piranha/tree/master/test-resources) ?

**Comment by ketkarameya** [comment]
Yes. For instance, you can update [this](https://github.com/uber/piranha/tree/master/test-resources/kt/file_scoped_chain_rules)

**Comment by JamesBondOOO7** [comment]
I was just running the `cargo test` command and saw that 1 test is failing.

```
test tests::test_piranha_python::test_delete_modify_str_literal_from_list_via_cli ... FAILED
```

I think this test is not related to the refactoring issue.

**Comment by ketkarameya** [comment]
aah! yea. You need to build and test. 
`cargo build && cargo test`

**Comment by JamesBondOOO7** [comment]
Yeah, it works 😅😅!!

**Comment by ketkarameya** [comment]
oh boy! You are going to have to rebase your changes :| I guess you will have to rename the macro now.

**Comment by JamesBondOOO7** [comment]
I tried to sync and ended up closing the PR.
Will try it again.

**Comment by ketkarameya** [comment]
@JamesBondOOO7  ow ! I think i know what happened. You were developing on the `master` of your fork . 
I would recommend creating a branch (of your fork), and making a PR for this branch against `uber:master`. 
This will make rebasing code easy. (Also I know, a lot has changes in the last commit) We introduced some macros for the DSL, so you might have to also rename the `constraint!` macro appropritaely. :| I think you just picked the task at a tumultous time. :P

**Pull request closed without merging**